README.md
# jssm-viz-cli
Command line visualizer for [fsl machines](https://fsl.tools/), producing flowcharts in SVG, PNG, DOT, or other formats.

Run the machines with [jssm](https://github.com/StoneCypher/jssm), or visualize within code (instead of at the CLI) with [jssm-viz](https://github.com/StoneCypher/jssm-viz-cli).

[Try it online](https://stonecypher.github.io/jssm-viz-demo/graph_explorer.html)

See the [fsl.tools](https://fsl.tools/) website 😊





<br/><br/>

## Getting started

If you want to use this as a CLI in general, install globally, which creates a command-line wrapper:

```bash
npm install -g jssm-viz-cli
```

And now you can do things like

```bash
jssm-viz -i traffic-light.fsl --png
```

or

```bash
jssm-viz -i traffic-light.fsl --png
```





<br/><br/>

## Scripting

If you'd prefer to use this as a CLI tool in a specific project, install it local to that project instead:

```bash
npm install --save jssm-viz
```

And now you can add things to your scripts block in `package.json` like

```javascript
{
  scripts: {
    "graph-fsl": "jssm-viz -i ./src/fsl/*.fsl --jpeg"
  }
}
```




<br/><br/>

### Install problems with `sharp.js`

The underlying `svg` converter, `sharp.js`,
[only exists for certain platforms](https://sharp.pixelplumbing.com/install#prebuilt-binaries).
This includes most common platforms.

1. If you're on Windows, make sure `node` is 64-bit.
1. On any platform, make sure you're on `node` versions 8, 12, or 13.
1. You'll need [python](https://www.python.org/downloads/)

If any of these are a problem, [the jssm-viz github action](https://github.com/StoneCypher/jssm-viz-action)
can do it for you with no hassle.





<br/><br/>

## Example

Assuming that in the file `traffic-light.fsl` we have the FSL machine

```fsl
machine_name: "Traffic light example";

Green 'next' => Yellow 'next' => Red 'next' => Green;
[Red Yellow Green] ~> Off -> Red;
```

We can run the command

```bash
jssm-viz -s traffic-light.fsl
```

You will produce the following image as SVG, in `traffic-light.svg`:

![](https://github.com/StoneCypher/jssm-viz-cli/blob/master/docs/traffic-light-example.svg)





<br/><br/>

## What if I want a PNG, JPEG, etc?

To write four images and two pieces of source to disk, run

```bash
jssm-viz -s traffic-light.fsl --svg --png --jpeg --webp --dot --tree
```

This will produce `traffic-light.svg`, `traffic-light.png`, `traffic-light.jpeg`, `traffic-light.webp`, and two special cases, `traffic-light.dot`, which is the intermediate stage that's run through [graphviz](https://www.graphviz.org/) to produce the final graph, and a parse tree in `jssm`'s internal format, in `json`.

If one or more of the format flags are provided, those are the formats which will be produced.  If none are provided, `svg` will be assumed.

There is also a `--jpg`, which produces JPEGs with the extension `.jpg`.  Internally it's treated like a different format, so, if you specify both `--jpg` and `--jpeg`, you'll get two images.

Images will be placed in the same directory as the source, unless otherwise specified.  This can be relevant if you're using globs to pick up multiple source files, by example.





<br/><br/>

## What if I want to set the render width?

For the non-vector non-source formats, it may be useful to control the render width.  `jssm-viz` will attempt to set a sensible default, but if you prefer to handle it yourself, this would create a 1234 px wide render of the machine `foo`:

```bash
jssm-viz -s foo.fsl --png --width 1234
```





<br/><br/>

## What if I want to render multiple machines?

The source flag `-s` takes a [glob](https://github.com/isaacs/node-glob#glob-primer), which can be a filename, but doesn't have to be.

By example, if you want to render all machines from the current directory downwards into both PNGs and SVGs, try

```bash
jssm-viz -s ./**/*.fsl --png --svg
```





<br/><br/>

## What if I want to control where the output goes?

Putting images where the source is isn't always desirable, and when you have a lot of machines, moving them after the render can be tedious.

`jssm-viz-cli` offers three output placement modes:

* In-place
* To directory
* To in-place directory
* To sourcenamed directory




<br/>

### In-place

```bash
jssm-viz -s foo.fsl --inplace
```

`In-place`, the default mode, produces an image with the source's filename and the relevant extension, in the same place that the source was found.





<br/>

### To directory

```bash
jssm-viz -s foo.fsl --todir ./renders
```

`To directory` allows you to specify a single directory that will contain all output. In this example, the directory `renders` would contain output (or be created if it didn't already exist.)

However, sometimes this causes filename conflicts, when different machines in different directories have the same filename.

Two modes help avoid those conflicts:





<br/>

### To in-place directory

```bash
jssm-viz -s foo.fsl --toinplacedir ./renders
```

`To in-place directory` will produce a path tree mimicing the original path tree under the target directory for matching `fsl` files, and place images there.  For example, `a/b/c/foo.fsl` aimed at `renders` will produce `renders/a/b/c/foo.svg`.





<br/>

### To sourcenamed directory

```bash
jssm-viz -s foo.fsl --tosourcenameddir ./renders
```

`To sourcenamed directory` will produce, for `a/b/c/foo.fsl`, `a--b--c--foo.svg`, which allows a flat directory to contain most patterns in a non-colliding easily [slug](https://www.npmjs.com/package/slug)ged way.




<br/><br/>

## What if I want this automated?

Github action coming ***Real Soon Now*** &trade;





<br/><br/>

## What does the SVG code look like?

This is the above example:

```xml
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN"
 "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<!-- Generated by graphviz version 2.40.1 (20161225.0304)
 -->
<!-- Title: G Pages: 1 -->
<svg width="125pt" height="260pt"
 viewBox="0.00 0.00 125.08 260.00" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
<g id="graph0" class="graph" transform="scale(1 1) rotate(0) translate(4 256)">
<title>G</title>
<polygon fill="#eeeeee" stroke="transparent" points="-4,4 -4,-256 121.0798,-256 121.0798,4 -4,4"/>
<!-- n0 -->
<g id="node1" class="node">
<title>n0</title>
<polygon fill="#ffffff" stroke="#000000" points="112.6045,-252 58.6045,-252 58.6045,-216 112.6045,-216 112.6045,-252"/>
<text text-anchor="middle" x="85.6045" y="-229.8" font-family="Times New Roman" font-size="14.00" fill="#000000">Green</text>
</g>
<!-- n1 -->
<g id="node2" class="node">
<title>n1</title>
<polygon fill="#ffffff" stroke="#000000" points="57.3139,-180 -.1049,-180 -.1049,-144 57.3139,-144 57.3139,-180"/>
<text text-anchor="middle" x="28.6045" y="-157.8" font-family="Times New Roman" font-size="14.00" fill="#000000">Yellow</text>
</g>
<!-- n0&#45;&gt;n1 -->
<g id="edge1" class="edge">
<title>n0&#45;&gt;n1</title>
<path fill="none" stroke="#333333" d="M71.221,-215.8314C64.6559,-207.5386 56.7538,-197.557 49.5467,-188.4533"/>
<polygon fill="#333333" stroke="#333333" points="52.1329,-186.0813 43.1817,-180.4133 46.6446,-190.4262 52.1329,-186.0813"/>
<text text-anchor="middle" x="66.0556" y="-210.4314" font-family="Open Sans" font-size="6.00" fill="#000000">next</text>
</g>
<!-- n3 -->
<g id="node4" class="node">
<title>n3</title>
<polygon fill="#ffffff" stroke="#000000" points="102.6045,-36 48.6045,-36 48.6045,0 102.6045,0 102.6045,-36"/>
<text text-anchor="middle" x="75.6045" y="-13.8" font-family="Times New Roman" font-size="14.00" fill="#000000">Off</text>
</g>
<!-- n0&#45;&gt;n3 -->
<g id="edge2" class="edge">
<title>n0&#45;&gt;n3</title>
<path fill="none" stroke="#bbbbbb" d="M93.8171,-215.6609C106.2756,-185.2499 126.9578,-122.7086 111.6045,-72 108.8348,-62.8522 103.9478,-53.8377 98.6509,-45.8725"/>
<polygon fill="none" stroke="#bbbbbb" points="98.6116,-45.818 91.8583,-43.2894 91.595,-36.0831 98.3482,-38.6116 98.6116,-45.818"/>
</g>
<!-- n2 -->
<g id="node3" class="node">
<title>n2</title>
<polygon fill="#ffffff" stroke="#000000" points="102.6045,-108 48.6045,-108 48.6045,-72 102.6045,-72 102.6045,-108"/>
<text text-anchor="middle" x="75.6045" y="-85.8" font-family="Times New Roman" font-size="14.00" fill="#000000">Red</text>
</g>
<!-- n1&#45;&gt;n2 -->
<g id="edge3" class="edge">
<title>n1&#45;&gt;n2</title>
<path fill="none" stroke="#333333" d="M40.4646,-143.8314C45.7674,-135.7079 52.1282,-125.9637 57.9718,-117.0118"/>
<polygon fill="#333333" stroke="#333333" points="61.0493,-118.7003 63.5847,-108.4133 55.1876,-114.8739 61.0493,-118.7003"/>
<text text-anchor="middle" x="35.2992" y="-138.4314" font-family="Open Sans" font-size="6.00" fill="#000000">next</text>
</g>
<!-- n1&#45;&gt;n3 -->
<g id="edge4" class="edge">
<title>n1&#45;&gt;n3</title>
<path fill="none" stroke="#bbbbbb" d="M28.3872,-143.8348C28.7191,-125.2351 30.7999,-95.7649 39.6045,-72 42.925,-63.0375 47.9978,-54.0854 53.2969,-46.121"/>
<polygon fill="none" stroke="#bbbbbb" points="53.3202,-46.0881 53.5306,-38.8801 60.2652,-36.302 60.0547,-43.5101 53.3202,-46.0881"/>
</g>
<!-- n2&#45;&gt;n0 -->
<g id="edge5" class="edge">
<title>n2&#45;&gt;n0</title>
<path fill="none" stroke="#333333" d="M76.8607,-108.0896C78.5609,-132.5727 81.6127,-176.5174 83.6233,-205.4713"/>
<polygon fill="#333333" stroke="#333333" points="80.1535,-206.0288 84.338,-215.7623 87.1367,-205.5438 80.1535,-206.0288"/>
<text text-anchor="middle" x="71.6953" y="-109.8896" font-family="Open Sans" font-size="6.00" fill="#000000">next</text>
</g>
<!-- n2&#45;&gt;n3 -->
<g id="edge6" class="edge">
<title>n2&#45;&gt;n3</title>
<path fill="none" stroke="#cccccc" d="M74.6045,-61.5726C74.6045,-57.3405 74.6045,-52.9622 74.6045,-48.6861"/>
<path fill="none" stroke="#777777" d="M76.6045,-61.5726C76.6045,-57.3405 76.6045,-52.9622 76.6045,-48.6861"/>
<polygon fill="none" stroke="#777777" points="72.1046,-61.8313 75.6045,-71.8314 79.1046,-61.8314 72.1046,-61.8313"/>
<polygon fill="none" stroke="#cccccc" points="75.6046,-48.4133 71.6045,-42.4133 75.6045,-36.4133 79.6045,-42.4132 75.6046,-48.4133"/>
</g>
</g>
</svg>
```

<br/><br/>

Don't mind the mess: this just got started Dec 28 of 2019.  Usable momentarily...


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

71 merges; Changlogging the last 10 commits; Full changelog at [CHANGELOG.long.md](CHANGELOG.long.md)



&nbsp;

&nbsp;

Published tags:







&nbsp;

&nbsp;

## [Untagged] - 6/24/2022 2:36:54 PM

Commit [bc013ddbf667d84fbd04177ca2394c7fbff7cff1](https://github.com/StoneCypher/jssm/commit/bc013ddbf667d84fbd04177ca2394c7fbff7cff1)

Author: `John Haugeland <stonecypher@gmail.com>`

  * this should only be testing ubuntu 18 in main path too




&nbsp;

&nbsp;

## [Untagged] - 6/24/2022 2:35:37 PM

Commit [d74646c9d441b8be1c8aec8a34623bfe467ed4e3](https://github.com/StoneCypher/jssm/commit/d74646c9d441b8be1c8aec8a34623bfe467ed4e3)

Author: `John Haugeland <stonecypher@gmail.com>`

  * this should only be testing ubuntu 18




&nbsp;

&nbsp;

## [Untagged] - 6/21/2022 6:45:10 PM

Commit [2a712877c9cd124ee38c2c8a6a3b16ae3c072fe5](https://github.com/StoneCypher/jssm/commit/2a712877c9cd124ee38c2c8a6a3b16ae3c072fe5)

Author: `John Haugeland <stonecypher@gmail.com>`

  * Prep for full chain auto-bump




&nbsp;

&nbsp;

## [Untagged] - 6/19/2022 9:51:16 PM

Commit [a69c417539d80e8336887ac7d8bcabf7234c9846](https://github.com/StoneCypher/jssm/commit/a69c417539d80e8336887ac7d8bcabf7234c9846)

Author: `John Haugeland <stonecypher@gmail.com>`

  * fonts are a mess in librsvg, aren't they




&nbsp;

&nbsp;

## [Untagged] - 6/19/2022 4:57:13 PM

Commit [3446ebddb5424499a8ab62810c36f63bf687769d](https://github.com/StoneCypher/jssm/commit/3446ebddb5424499a8ab62810c36f63bf687769d)

Author: `John Haugeland <stonecypher@gmail.com>`

  * hey, this is actually working already




&nbsp;

&nbsp;

## [Untagged] - 6/1/2022 4:38:33 PM

Commit [df447d707f0532fa6c769a6162a682c97f2dccd8](https://github.com/StoneCypher/jssm/commit/df447d707f0532fa6c769a6162a682c97f2dccd8)

Author: `dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>`

  * Bump sharp from 0.24.1 to 0.30.5
  * Bumps [sharp](https://github.com/lovell/sharp) from 0.24.1 to 0.30.5.
- [Release notes](https://github.com/lovell/sharp/releases)
- [Changelog](https://github.com/lovell/sharp/blob/main/docs/changelog.md)
- [Commits](https://github.com/lovell/sharp/compare/v0.24.1...v0.30.5)
  * ---
updated-dependencies:
- dependency-name: sharp
  dependency-type: direct:production
...
  * Signed-off-by: dependabot[bot] <support@github.com>




&nbsp;

&nbsp;

## [Untagged] - 3/26/2022 3:26:59 PM

Commit [0836cae05593d00ffb209d065add9dc496f9655d](https://github.com/StoneCypher/jssm/commit/0836cae05593d00ffb209d065add9dc496f9655d)

Author: `John Haugeland <stonecypher@gmail.com>`

Merges [e87e070, 8e86c45]

  * Merge pull request #29 from StoneCypher/dependabot/npm_and_yarn/simple-get-3.1.1
  * Bump simple-get from 3.1.0 to 3.1.1




&nbsp;

&nbsp;

## [Untagged] - 3/26/2022 12:29:30 PM

Commit [8e86c45bbd0d90074e069e2444412fdd47126fb1](https://github.com/StoneCypher/jssm/commit/8e86c45bbd0d90074e069e2444412fdd47126fb1)

Author: `dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>`

  * Bump simple-get from 3.1.0 to 3.1.1
  * Bumps [simple-get](https://github.com/feross/simple-get) from 3.1.0 to 3.1.1.
- [Release notes](https://github.com/feross/simple-get/releases)
- [Commits](https://github.com/feross/simple-get/compare/v3.1.0...v3.1.1)
  * ---
updated-dependencies:
- dependency-name: simple-get
  dependency-type: indirect
...
  * Signed-off-by: dependabot[bot] <support@github.com>




&nbsp;

&nbsp;

## [Untagged] - 3/26/2022 12:29:07 PM

Commit [e87e070325b625b68a57ae7fbdb5d7c58e5b5f93](https://github.com/StoneCypher/jssm/commit/e87e070325b625b68a57ae7fbdb5d7c58e5b5f93)

Author: `John Haugeland <stonecypher@gmail.com>`

Merges [42e4f46, 9e03179]

  * Merge pull request #28 from StoneCypher/dependabot/npm_and_yarn/minimist-1.2.6
  * Bump minimist from 1.2.5 to 1.2.6




&nbsp;

&nbsp;

## [Untagged] - 3/25/2022 6:35:28 PM

Commit [9e031795bc7f7c36375d73d5e98986be75f40f6e](https://github.com/StoneCypher/jssm/commit/9e031795bc7f7c36375d73d5e98986be75f40f6e)

Author: `dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>`

  * Bump minimist from 1.2.5 to 1.2.6
  * Bumps [minimist](https://github.com/substack/minimist) from 1.2.5 to 1.2.6.
- [Release notes](https://github.com/substack/minimist/releases)
- [Commits](https://github.com/substack/minimist/compare/1.2.5...1.2.6)
  * ---
updated-dependencies:
- dependency-name: minimist
  dependency-type: indirect
...
  * Signed-off-by: dependabot[bot] <support@github.com>

rollup.config.js

import nodeResolve from '@rollup/plugin-node-resolve';
import commonjs    from '@rollup/plugin-commonjs';
import json        from '@rollup/plugin-json';

import typescript  from 'rollup-plugin-typescript2';
import ignore      from 'rollup-plugin-ignore';

const pkg = require('./package');





const cjsconfig = {

  input     : 'src/ts/cli.ts',

  output    : {
    file      : 'build/cli.js',
    format    : 'cjs',
    name      : 'bin',
    banner    : '#!/usr/bin/env node\n'
  },

  plugins   : [

    json(),

    ignore(['fs', 'path', 'crypto']),

    typescript(),

    commonjs({
      include: 'node_modules/**'
    }),

    nodeResolve({
      mainFields     : [ 'module', 'main' ],
      browser        : true,
      extensions     : [ '.js', '.json', '.ts' ],
      preferBuiltins : false
    })

  ]

};






export default cjsconfig;


src/ts/cli.ts

// shebang is in rollup banner; see bin-rollup.config.js





const fs    = require('fs'),
      path  = require('path');

const sharp = require('sharp'),
      glob  = require('glob'),
      app   = require('commander'),
      a2c   = require('ansi-256-colors');


import { version }           from '../../package.json';
import { file_type }         from './types';

import {
  fsl_to_svg_string,
  fsl_to_dot
} from 'jssm-viz';





const DEFAULT_SCALE = 3;

const rasterFormats = ['png', 'jpg', 'jpeg', 'webp'],
      otherFormats  = ['svg', 'tree', 'dot'],
      imgFormats    = [].concat(rasterFormats, otherFormats),
      dirsOn        = ['inplace', 'todir', 'toinplacedir', 'tosourcenameddir', 'topipe'],
      noise         = ['debug', 'verbose', 'quiet', 'silent'];

type imageFormatsAsTuple = typeof imgFormats; // see https://stackoverflow.com/a/45486495/763127
type imgFormat           = imageFormatsAsTuple[number];





const at      = (s, i)              => parseInt(`${s}`.padStart(3, '0')[i]),
      fg_from = colstr              => (colstr === false)? '' : a2c.fg.getRgb(at(colstr, 0), at(colstr, 1), at(colstr, 2)),
      cx      = (color, text)       => fg_from(color) + text + a2c.reset,
      col     = (num, text)         => app.color? `${cx(num, text)}` : text,
      cols    = (nt_pairs)          => nt_pairs.map( pair => col(pair[0], pair[1]) ).join(''),
      if_text = (label, text, tcol) => text? (label + col(tcol, text)) : '';

const error_text = text => if_text(col(501, 'Error: '), text, 412) + `${col(441, " (see ")}${col(141, "jssm-viz --help")}${col(441, " for details)")}`,
      debug_text = text => if_text(col(113, 'Debug: '), text, 12),
      quiet_text = text => if_text(col(131, 'Quiet: '), text, 21);

let scale, height, width;





const render_message = fname =>
  app.color
    ? `${col(222, ' - ')}${col(123, "Rendering")} ${col(135, fname)}`
    : ` - Rendering ${fname}`;

const render_result = fname =>
  app.color
    ? `${col(222, '   - ')}${col(123, "Output")} ${col(135, fname)}`
    : `   - Output ${fname}`;





const log_if      = (text, clause) => clause? process.stdout.write(text + '\n') : true,

      debug_log   = text => log_if(debug_text(text), ['debug'                    ].includes(app.noise_level)),
      verbose_log = text => log_if(           text , ['debug', 'verbose'         ].includes(app.noise_level)),
      quiet_log   = text => log_if(quiet_text(text), ['debug', 'verbose', 'quiet'].includes(app.noise_level));





function accumulateSource(value, previous) {
  return previous.concat([value]);
}





app

  .version(version)

  .option('-i, --input <glob>',       'the input source file, as a glob, such as foo.fsl or ./**/*.fsl')
  .option('-o, --output <string>',    'the output filename, as a string.  Omit and the input fname will be used')
  .option('-O, --outputDir <string>', 'the output directory, as a string; otherwise predict the filenames as normal')
  .option('-0, --outputConsole',      'output to standard output / pipe 0 / cout; implies -q otherwise')
  .option('-N, --nestedDir <string>', 'the output directory, retaining subdir relative to here; otherwise predict')
  .option('-s, --source <string>',    'provide source as a string', undefined, accumulateSource, [])

  .option('-w, --width <integer>',    'set raster render width, in pixels (exclusive of height or scale)')
  .option('-h, --height <integer>',   'set raster render height, in pixels (exclusive of width or scale)')
  .option('-x, --scale <number>',     `set raster scale, in pixels (exclusive of width or height, default ${DEFAULT_SCALE})`)

  .option('-d, --debug',              'log extensively to console')
  .option('-v, --verbose',            'log to console normally (default)')
  .option('-q, --quiet',              'only log to console on error')
  .option('-z, --silent',             'do not log to console at all')

  .option('-c, --color',              'use console color (default)')
  .option('-n, --nocolor',            'do not use console color')

  .option('-S, --svg',                'produce output in SVG format (default if no formats specified)')
  .option('-P, --png',                'produce output in PNG format')
  .option('-J, --jpg',                'produce output in JPEG format, with a .jpg extension')
  .option('-E, --jpeg',               'produce output in JPEG format, with a .jpeg extension')
  .option('-W, --webp',               'produce output in WEBP format')
  .option('-T, --tree',               'produce output in JSSM\'s internal parse tree format, with a .tree extension')
  .option('-D, --dot',                'produce output in GraphViz\'s DOT format')

app.parse(process.argv);





function english_list(list) {

  if (list.length === 0) { return ""; }
  if (list.length === 1) { return `${list[0]}`; }
  if (list.length === 2) { return `${list[0]} and ${list[1]}`; }

  const [ last, ... r_front ] = list.reverse(),
        front                 = r_front.reverse();

  return `${front.join(', ')}, and ${last}`;

}





const present_on_app = (test_items) =>
  test_items.filter(ti => app[ti] !== undefined);





function validate_args() {

  const colors = present_on_app(['color', 'nocolor']);
  if (colors.length > 1) {
    console.log(error_text(`${english_list(['color', 'nocolor'])} are mutually exclusive.  Please choose at most one.`));
    process.exit(1);
  }
  if (colors.length === 0) {
    app.color = true;
  }

  const sizes = present_on_app(['height', 'width', 'scale']);
  if (sizes.length > 1) {
    console.log(error_text(`${english_list(['height', 'width', 'scale'])} are mutually exclusive.  Please choose at most one.`));
    process.exit(1);
  }

  if (sizes.length === 0) { scale  = DEFAULT_SCALE; }
  if (app.scale)          { scale  = app.scale;     }
  if (app.height)         { height = app.height;    }
  if (app.width)          { width  = app.width;     }

  const noises = present_on_app(noise);
  if (noises.length > 1) {
    console.log(`${english_list(noises)} are mutually exclusive.  Please choose at most one.`);
    process.exit(1);
  } else if (noises.length === 1) {
    app.noise_level = noises[0];
    debug_log(`Noise level: ${app.noise_level}`);
  } else {
    debug_log(`No noise level specified; defaulting to verbose`);
    app.noise_level = 'verbose';
  }

  const if_colored = colorer => app.color? colorer : ( x => x );

  if (!(app.input)) {
    console.log(error_text("must specify a source file or source glob with -i"));
    process.exit(1);
  }

  const uDirsOn = present_on_app(dirsOn);
  if (uDirsOn.length > 1) {
    console.log(error_text(`${english_list(dirsOn)} are mutually exclusive.  Please choose at most one.`));
    process.exit(1);
  }

  if (uDirsOn.length === 0) {
    debug_log(`No directory strategy specified; defaulting to inplace`);
    app.inplace = true;
  } else {
    debug_log(`Directory strategy: ${uDirsOn[0]}`);
  }

  if (present_on_app(imgFormats).length === 0) {
    debug_log(`No image format(s) specified; defaulting to svg`);
    app.svg = true;
  }

  debug_log(''); // on debug only emit a newline before the work output

}





function outputTarget(origFname, kind) {
  return origFname + '.' + kind; // lol TODO FIXME
}





function pull_size_from_svg(svg: string) {

  const inter      = `${svg}`.split('<svg width="'),
        after      = inter[1].split('pt" height="'),
        final      = after[1].split('pt"'),

        svg_width  = Number(after[0]),
        svg_height = Number(final[0]);

  return { svg_width, svg_height };

}





async function sharp_raster(sbuf, target_fmt: string, options?: object | undefined) {

  process.env.FONTCONFIG_PATH= `/c/Users/john/projects/jssm-viz-cli/src/fonts`;
//process.env.FONTCONFIG_PATH= `${__filename}/fonts`;

  const u_opt = (options === undefined)
                  ? {}
                  : options;

  if (!( imgFormats.includes(target_fmt) )) {
    throw new TypeError(`unknown format: ${target_fmt}`);
  }

  if (['dot', 'tree', 'svg'].includes(target_fmt)) {
    throw new TypeError(`not a raster format: ${target_fmt}`);
  }

  const { svg_width, svg_height } = pull_size_from_svg(sbuf);

  let final_width  = svg_width,
      final_height = svg_height;

  if (scale !== undefined) {
    final_height = Math.ceil(final_height * scale);
    final_width  = Math.ceil(final_width  * scale);
  } else if (height !== undefined) {
    final_height = Math.ceil(height);
    final_width  = Math.ceil((height / svg_height) * width);
  } else if (width !== undefined) {
    final_width  = Math.ceil(width);
    final_height = Math.ceil((width / svg_width) * height);
  }

  const ufmt      = (target_fmt === 'jpg')? 'jpeg' : target_fmt,
        sharp_buf = await sharp(sbuf)
                            .resize({
                              width  : final_width,
                              height : final_height,
                              fit    : 'contain'
                            });

  const buf       = sharp_buf[ufmt]().toBuffer();

  return buf;

}





// const to_inject = `
//   <style type="text/css">
//     @font-face {
//       font-family : Open Sans;
//       src         : '${__dirname}/fonts/OpenSans-Regular.ttf';
//     }
//   </style>
// `;

// const to_inject = `
//   <style type="text/css">
//     @font-face {
//       font-family : 'Open Sans';
//       src         : url('../../dist/js/fonts/OpenSans-Regular.ttf');
//     }
//   </style>
// `;

// const to_inject = `
//   <style type="text/css">
//     @font-face {
//       font-family : 'Open Sans';
//       src         : url('${path.normalize(__dirname).replace(/\\/g,'\\\\')}/fonts/OpenSans-Regular.ttf');
//     }
//   </style>
// `;

// const to_inject = `
//   <style type="text/css">
//     @font-face {
//       font-family : 'Open Sans';
//       src         : url('C:\\Users\\john\\projects\\jssm-viz-cli\\dist\\js\\fonts\\OpenSans-Regular.ttf');
//     }
//   </style>
// `;

// const to_inject = `
//   <style type="text/css">
//     @font-face {
//       font-family : Open Sans;
//       src         : url('../../dist/js/fonts/OpenSans-Regular.ttf');
//     }
//   </style>
// `;

// const to_inject = `
//   <style type="text/css">
//     @font-face {
//       font-family : Open Sans;
//       src         : url('./OpenSans-Regular.ttf');
//     }
//   </style>
// `;

// const to_inject = `
//   <style type="text/css">
//     @font-face {
//       font-family : Open Sans;
//       src         : url('.\\\\OpenSans-Regular.ttf');
//     }
//   </style>
// `;

// const to_inject = `
//   <style type="text/css">
//     @font-face {
//       font-family : Open Sans;
//       src         : url('OpenSans-Regular.ttf');
//     }
//   </style>
// `;

// const to_inject = `
//   <style type="text/css">
//     @font-face {
//       font-family : Open Sans;
//       src         : OpenSans-Regular.ttf;
//     }
//   </style>
// `;

const to_inject = `
  <style type="text/css">
    @font-face {
      font-family : Bonk;
      src         : url('./Impact.ttf');
    }
  </style>
`;


const font_split_token = 'http://www.w3.org/1999/xlink">';

function inject_css(svg_str) {

  const parts = svg_str.split(font_split_token);
  parts[1]    = to_inject + parts[1];
  const whole = parts.join(font_split_token);

//  return whole;
//  return whole.replace(/font-family="Open Sans"/g, `font-family="'Open Sans'"`);
  // return whole.replace(/font-family="Open Sans"/g, `font-family="Impact"`);
  return whole.replace(/font-family="Open Sans"/g, `font-family="Open Sans, Helvetica Neue, Helvetica, Arial, sans-serif"`);

}





// TODO typeify the formats
async function write_sharp_raster(fname: string, svg_str: string, fmt: string) {

  const buf = await sharp_raster( Buffer.from( svg_str ), fmt );
  fs.writeFileSync(outputTarget(fname, fmt), buf);
  verbose_log(render_result(outputTarget(fname, fmt)));

}





async function output({ fname, data }) {

  verbose_log(render_message(fname));

  const svgo = await fsl_to_svg_string(data),
        svg  = inject_css(svgo),
        dot  = await fsl_to_dot(data),
        sbuf = Buffer.from(svg);

  let written = 0;

  if (app.svg) {
    fs.writeFileSync(outputTarget(fname, 'svg'), svg);
    verbose_log(render_result(outputTarget(fname, 'svg')));
    ++written;
  }

  if (app.dot) {
    fs.writeFileSync(outputTarget(fname, 'dot'), dot);
    verbose_log(render_result(outputTarget(fname, 'dot')));
    ++written;
  }

  // TODO handle tree

  const handles = rasterFormats.map( async thisFmt => {

    if (app[thisFmt]) {
      await write_sharp_raster( fname, svg, thisFmt );
      ++written;
    }

  } );

  await Promise.all(handles);

  if (written === 0) {
    // TODO FIXME there should be error handling here once the actual features
    // are filled out
  }

}





async function run() {

  validate_args();
  console.log(`${col(12, 'jssm-viz cli ')}${col(111, `version ${version}`)}`);

  console.log('');
  verbose_log(`${col(345, "jssm-viz: ")}${col(135, 'targetting ')}${col(24, english_list(present_on_app(imgFormats)))}`);

  const files = glob.sync(app.input);

  if (files.length === 0) {
    console.log(error_text(`no files found matching source glob ${col(441, app.input)}`))
  }

  const o_promises = files.map(fname => ({fname, data: `${fs.readFileSync(fname)}`}))
                          .map(output);

  await Promise.all(o_promises);

  console.log('');
  verbose_log(`${col(345, "jssm-viz: ")}${col(135, `finished successfully`)}`);

}





run();
